src/core/qgsproperty.h

```diff
@@ -0,0 +1,929 @@
+/***************************************************************************
+     qgsproperty.h
+     -------------
+    Date                 : April 2015
+    Copyright            : (C) 2015 by Nyall Dawson
+    Email                : nyall dot dawson at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+#ifndef QGSPROPERTY_H
+#define QGSPROPERTY_H
+
+#include "qgsexpression.h"
+#include "qgsexpressioncontext.h"
+#include <QVariant>
+#include <QHash>
+#include <QString>
+#include <QStringList>
+#include <QDomElement>
+#include <QDomDocument>
+#include <QColor>
+
+class QgsPropertyTransformer;
+class QgsVectorColorRampV2;
+
+/** \ingroup core
+ * \class QgsAbstractProperty
+ * \brief Abstract base class for properties.
+ *
+ * QgsAbstractProperty objects are used for storing properties for objects, which can then be evaluated using
+ * custom logic by evaluating them against a supplied QgsExpressionContext. Multiple QgsAbstractProperty objects
+ * can be grouped using a QgsPropertyCollection for easier bulk storage, retrieval and evaluation.
+ * \note Added in version 2.16
+ */
+
+class CORE_EXPORT QgsAbstractProperty
+{
+  public:
+
+    //! Property types
+    enum Type
+    {
+      StaticProperty, /*!< Static property (QgsStaticProperty) */
+      FieldBasedProperty, /*!< Field based property (QgsFieldBasedProperty) */
+      ExpressionBasedProperty, /*!< Expression based property (QgsExpressionBasedProperty) */
+    };
+
+    /** Factory method for creating a new property of the specified type.
+     * @param type property type to create
+     */
+    static QgsAbstractProperty* create( Type type );
+
+    /** Constructor for a QgsAbstractProperty.
+     * @param active whether the property is initially active
+     */
+    QgsAbstractProperty( bool active = true );
+
+    //! Copy constructor
+    QgsAbstractProperty( const QgsAbstractProperty& other );
+
+    QgsAbstractProperty& operator=( const QgsAbstractProperty& other );
+
+    virtual ~QgsAbstractProperty();
+
+    /** Returns the property type.
+     */
+    virtual Type propertyType() const = 0;
+
+    /** Returns a clone of the property.
+     */
+    virtual QgsAbstractProperty* clone() = 0;
```

Comment:
This will force us to always use properties with pointers.
Would it be possible to have them implicitly shared and not having to care for the memory management? I think it would make usage much more comfortable.